test/test_nn.py

```diff
@@ -5091,86 +5092,170 @@ def test_batchnorm_nhwc_cuda(self):
-        inp2 = inp1.contiguous(memory_format=torch.channels_last)
-        out1 = model(inp1)
-        out2 = model(inp2)
-        self.assertEqual(out1, out2)
+    @parametrize_test("mode", ["train", "inference"], name_fn=lambda x: x)
+    @parametrize_test(
+        # test verifies cudnn/miopen batchnorm with the reference backend or memory format
+        # memory_format - one of ("NCHW", NHWC")
+        # ref_backend - one of ("cpu", "native", "NCHW", "NHWC")
+        #   "cpu"    - cpu backend with the same memory_format will be used as reference
+        #   "native" - native backend (`with torch.backends.cudnn.flags(enabled=False)`)
+        #              with the same memory_format will be used
+        #   "NCHW" or "NHWC" - the same backend will be used but another memory format
+        # mixed - True or False. Mixed batchnorm mode where inputs are 16-bit and batchnorm is fp32
+        #
+        "memory_format,ref_backend,mixed,dtype",
+        [
+            ("NCHW", "cpu", False, torch.float),
+            ("NCHW", "cpu", True, torch.half),
+            ("NCHW", "cpu", True, torch.bfloat16),
+
+            ("NCHW", "native", False, torch.float),
+            ("NCHW", "native", True, torch.half),
+            # this config failed for train and passed for inference
+            # subtest(("NCHW", "native", True, torch.bfloat16), decorators=[unittest.expectedFailure]),
+
+            ("NHWC", "cpu", False, torch.float),
+            ("NHWC", "cpu", True, torch.half),
+            ("NHWC", "cpu", True, torch.bfloat16),
+
+            ("NHWC", "native", False, torch.float),
+            ("NHWC", "native", True, torch.half),
+            ("NHWC", "native", True, torch.bfloat16),
+
+            ("NHWC", "NCHW", False, torch.float),
+            ("NHWC", "NCHW", True, torch.half),
+            ("NHWC", "NCHW", True, torch.bfloat16),
+        ],
+        name_fn=lambda f, b, m, t: f"{f}_vs_{b}{'_mixed' if m else ''}_{dtype_name(t)}"
+    )
+    def test_batchnorm(self, mode, memory_format, ref_backend, mixed, dtype):
+        def _create_tensor(size, memory_format, dtype, device):
+            t = torch.empty(size=size, memory_format=memory_format, dtype=dtype, device=device)
+            t = t.random_(1, 10)
+            return t
+
+        def _get_ref_device(backend: str | None = None, device: str | None = None):
+            if backend is None or backend == "native":
+                return "cuda"
+            if backend == "cpu":
+                return "cpu"
+            if backend in ("NHWC", "NCHW"):
+                if device is not None:
+                    return device
+                else:
+                    raise ValueError("device required for backend ('NHWC', 'NCHW')")
+            else:
+                raise ValueError("Unknown backend")
+
+        def _get_backend_memory_format(backend: str, memory_format: str | None = None):
+            if backend == "NHWC":
```

Comment:
```suggestion
            # If backend specifies the memory format, return it, otherwise look at memory_format arg
            if backend == "NHWC":
```